File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {
    const n = grid.length

    if (grid[0][0] === 1 || grid[n-1][n-1] === 1) return -1

    const directions = [[1,1],[1,-1],[-1,-1],[-1,1],[1,0],[-1,0],[0,1],[0,-1]]

    const visited = new Set()

    const queue = [[0, 0, 1]]

    visited.add(createPositionKey(0, 0))

    while (queue.length > 0) {
        const [row, col, pathLength] = queue.shift()
        
        if (row === n - 1 && col === n - 1) {
            return pathLength
        }

        for (const [dx, dy] of directions) {
            const newRow = row + dx
            const newCol = col + dy

            if (isValidAndUnvisited(grid, newRow, newCol, visited)) {
                visited.add(createPositionKey(newRow, newCol))
                queue.push([newRow, newCol, pathLength + 1])
            }
        }
    }
    return -1
};

function createPositionKey(row, col) {
    return `${row},${col}`
}

function isValidAndUnvisited(grid, row, col, visited) {
    if (row < 0 || row >= grid.length || col < 0 || col >= grid.length) return false

    return grid[row][col] === 0 && !visited.has(createPositionKey(row, col))
}

File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        n = len(grid)
        m = len(grid[0])

            # Check if start or end is blocked
        if grid[0][0] == 1 or grid[n-1][m-1] == 1:
            return -1

        directions = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1),           (0, 1),
        (1, -1),  (1, 0),  (1, 1)]

        grid[0][0] = 1

        queue = deque([(0, 0, 1)])

        while queue:
            row, col, steps = queue.popleft()

            if row == n - 1 and col == m - 1:
                return steps

            for dr, dc in directions:
                new_row = row + dr
                new_col = col + dc

                if new_row < 0 or new_row >= n:
                    continue

                if new_col < 0 or new_col >= m:
                    continue

                if grid[new_row][new_col] == 1:
                    continue

                grid[new_row][new_col] = 1

                queue.append((new_row, new_col, steps + 1))

        return -1

